core/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -392,11 +392,118 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
+        ElasticsearchException cause = null;
+        Map<String, List<String>> metadata = new HashMap<>();
+        Map<String, List<String>> headers = new HashMap<>();
+
+        XContentParser.Token token = parser.currentToken();
+        String currentFieldName = parser.currentName();
+        do {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                token = parser.nextToken();
+            }
+
+            if (token != null && token.isValue()) {
+                if (TYPE.equals(currentFieldName)) {
+                    type = parser.text();
+                } else if (REASON.equals(currentFieldName)) {
+                    reason = parser.text();
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    metadata.put(currentFieldName, Collections.singletonList(parser.text()));
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (CAUSED_BY.equals(currentFieldName)) {
+                    cause = fromXContent(parser);
+                } else if (HEADER.equals(currentFieldName)) {
+                    Map<String, Object> headerMap = parser.map();
+                    if (headerMap != null) {
+                        for (Map.Entry<String, Object> header : headerMap.entrySet()) {
+                            List<String> values = headers.getOrDefault(header.getKey(), new ArrayList<>());
+                            if (header.getValue() instanceof Iterable) {
+                                for (Object value : (Iterable) header.getValue()) {
+                                    values.add(String.valueOf(value));
+                                }
+                            } else {
+                                values.add(String.valueOf(header.getValue()));
+                            }
+                            headers.put(header.getKey(), values);
+                        }
+                    }
+                } else {
+                    // Any additional metadata object added by the metadataToXContent method is ignored
+                    // and skipped, so that the parser does not fail on unknown fields. The parser only
+                    // support metadata key-pairs and metadata arrays of values.
+                    parser.skipChildren();
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                // Parse the array and add each item to the corresponding list of metadata.
+                // Arrays of objects are not supported yet and just ignored and skipped.
+                List<String> values = new ArrayList<>();
+                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                    if (token.isValue() || token == XContentParser.Token.VALUE_NULL) {
```

Comment:
It supports null because headers/metadata/resources can contain null values and those null values will be present in the printed xcontent. So I think we have to parse them and it doesn't add extra complexity.



> question: what happens with fields that are not strings? do they become strings?

`parser.textOrNull()` takes care to retrieve a textual representation of the value but we might lose some precision.